src/main/java/net/neoforged/neoforge/common/world/ModifiableBiomeInfo.java

```diff
@@ -79,6 +83,18 @@ public void applyBiomeModifiers(final Holder<Biome> biome, final List<BiomeModif
             }
         }
         this.modifiedBiomeInfo = builder.build();
+        return !equivalent(originalBiomeInfo, modifiedBiomeInfo);
+    }
+
+    private boolean equivalent(BiomeInfo original, BiomeInfo modified) {
+        if (!original.climateSettings().equals(modified.climateSettings())) {
+            return false;
+        }
+        var oEffects = original.effects();
+        var mEffects = modified.effects();
+        JsonElement oEffectsJson = BiomeSpecialEffects.CODEC.encodeStart(JsonOps.INSTANCE, oEffects).result().orElse(null);
+        JsonElement mEffectsJson = BiomeSpecialEffects.CODEC.encodeStart(JsonOps.INSTANCE, mEffects).result().orElse(null);
+        return oEffectsJson != null && oEffectsJson.equals(mEffectsJson);
     }
```

Comment:
`Objects.equals()` is a nice shortcut for that.

Reply:
It's not, actually. We want to return `false` if either is null, even if _both_ are null, because that means serialization failed and we should assume they're different to be safe.

Reply:
That said, I should probably use a different DynamicOps here to be safe and grab a RegistryOps if I can get at one.